File: backend/app/schemas/safety_stock.py
from datetime import datetime
from pydantic import BaseModel, Field, field_validator

from app.schemas._limits import MAX_NUMERIC_10_2


class SafetyStockLineIn(BaseModel):
    product_id: int
    quantity: float = Field(gt=0, le=MAX_NUMERIC_10_2, description="Must be greater than zero.")

    @field_validator("quantity")
    @classmethod
    def round_quantity(cls, v: float) -> float:
        return round(v, 2)


class SafetyStockSave(BaseModel):
    # The branch's complete safety-stock list — replaces whatever was
    # saved before. A product left out is cleared; an empty list clears all.
    lines: list[SafetyStockLineIn]


class SafetyStockOut(BaseModel):
    # product_id -> quantity; a missing product_id means none set.
    quantities: dict[int, float]
    updated_at: datetime | None = None
    updated_by_username: str | None = None

File: frontend/src/api/safetyStock.ts
import { apiFetch } from "./client";

export type SafetyStock = {
  // product_id -> quantity; a missing product_id means none set.
  quantities: Record<number, number>;
  updated_at: string | null;
  updated_by_username: string | null;
};

export function fetchMySafetyStock(): Promise<SafetyStock> {
  return apiFetch("/safety-stock/mine");
}

// Replaces the branch's whole list — an empty `lines` clears everything.
export function saveMySafetyStock(lines: { product_id: number; quantity: number }[]): Promise<SafetyStock> {
  return apiFetch("/safety-stock/mine", {
    method: "PUT",
    body: JSON.stringify({ lines }),
  });
}
